crates/ark/src/lsp/main_loop.rs

```diff
@@ -770,44 +802,94 @@ async fn process_indexer_queue(mut rx: mpsc::UnboundedReceiver<IndexerQueueTask>
-        .into_iter()
-        .map(|task| match &task {
-            IndexerTask::Update { uri, .. } => (Some(uri.clone()), task),
-            IndexerTask::Start { .. } => (None, task),
-        })
-        .collect();
+    tracing::trace!(
+        "Processing {n} indexer tasks ({summary})",
+        n = batch.len(),
+        summary = summarize_indexer_task(&batch)
+    );
+
+    let mut updates: HashMap<PathBuf, IndexerTask> = HashMap::new();
+
+    fn flush_updates(
+        updates: &mut HashMap<PathBuf, IndexerTask>,
+    ) -> Vec<tokio::task::JoinHandle<()>> {
+        tracing::trace!("Flushing {n} update tasks", n = updates.len());
+
+        let mut handles = Vec::new();
+        for (path, task) in updates.drain() {
+            handles.push(tokio::task::spawn_blocking(move || {
+                if let IndexerTask::Update { document, .. } = task {
+                    if let Err(err) = indexer::update(&document, &path) {
+                        log::warn!("Indexer update failed: {err}");
+                    }
+                }
+            }));
+        }
+        handles
+    }
 
-    let mut handles = Vec::new();
+    let to_path_buf = |uri: &url::Url| {
+        uri.to_file_path()
+            .map_err(|_| anyhow!("Failed to convert URI '{uri}' to file path"))
+    };
 
-    for (_, task) in batch {
-        handles.push(tokio::task::spawn_blocking(move || match task {
-            IndexerTask::Start { folders } => {
-                indexer::start(folders);
-            },
-            IndexerTask::Update { document, uri } => {
-                let result = if let Ok(path) = uri.to_file_path() {
-                    indexer::update(&document, &path)
-                } else {
-                    Err(anyhow!("Failed to convert URI to file path: {uri}"))
-                };
-                if let Err(err) = result {
-                    log::error!("Indexer update failed: {err}");
-                }
-            },
-        }));
+    for task in batch {
+        let result: anyhow::Result<()> = (|| async {
+            match &task {
+                IndexerTask::Create { uri } => {
+                    let path = to_path_buf(uri)?;
+                    indexer::create(&path)?;
+                },
+
+                IndexerTask::Update { uri, .. } => {
+                    let path = to_path_buf(uri)?;
+                    updates.insert(path, task);
+                },
+
+                IndexerTask::Delete { uri } => {
+                    let handles = flush_updates(&mut updates);
+                    for handle in handles {
+                        let _ = handle.await;
+                    }
+                    let path = to_path_buf(uri)?;
+                    indexer::delete(&path)?;
+                },
+
+                IndexerTask::Rename {
+                    uri: old_uri,
+                    new: new_uri,
+                } => {
+                    let handles = flush_updates(&mut updates);
```

Comment:
I'm very confused as to why we batch the `Update` tasks. I figure we would just process them in order. Batching them like this (and across different URIs too?) feels pretty complicated?

Is the idea to collapse multiple `IndexerTask::Update` calls on the same document together into just one run of `indexer::update()`?

Having to manage the timing of 3 separate calls to `flush_updates()` doesn't feel worth it for that. Is there evidence of a performance issue here?

Reply:
Now a simpler queue without any batching, we'll revisit if performance is lacking.